File: SemanticSearch/Java/Web/SpringBoot/src/main/java/com/portfolio/semanticsearch/vectorstore/ChromaDbVectorStore.java
package com.portfolio.semanticsearch.vectorstore;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class ChromaDbVectorStore implements VectorStoreAdapter {
    private final RestTemplate restTemplate;
    private final ObjectMapper mapper;
    private final String chromaUrl;
    private final int dimension;
    private final String collection;
    private String collectionId;

    public ChromaDbVectorStore(String chromaUrl, int dimension, String collection) {
        this(new RestTemplate(), new ObjectMapper(), chromaUrl, dimension, collection);
    }

    public ChromaDbVectorStore(RestTemplate restTemplate, ObjectMapper mapper, String chromaUrl, int dimension, String collection) {
        this.restTemplate = restTemplate;
        this.mapper = mapper;
        this.chromaUrl = chromaUrl;
        this.dimension = dimension;
        this.collection = collection;
    }

    @Override
    public void connect() {
        try {
            ResponseEntity<String> res = restTemplate.getForEntity(chromaUrl + "/api/v2/collections", String.class);
            List<Map<String, Object>> existing = new ArrayList<>();
            try {
                existing = mapper.readValue(res.getBody(), mapper.getTypeFactory().constructCollectionType(List.class, Map.class));
            } catch (Exception ignored) {
            }
            String name = collection;
            String id = null;
            for (Map<String, Object> c : existing) {
                if (name.equals(c.get("name"))) {
                    id = (String) c.get("id");
                    break;
                }
            }
            if (id == null) {
                Map<String, Object> create = new LinkedHashMap<>();
                create.put("name", name);
                create.put("metadata", Map.of("hnsw:space", "cosine"));
                create.put("getOrCreate", true);
                HttpHeaders headers = new HttpHeaders();
                headers.setContentType(MediaType.APPLICATION_JSON);
                ResponseEntity<Map> created = restTemplate.postForEntity(chromaUrl + "/api/v2/collections", new HttpEntity<>(create, headers), Map.class);
                Map<String, Object> body = created.getBody();
                if (body != null) {
                    id = (String) body.get("id");
                    name = (String) body.getOrDefault("name", name);
                }
            }
            this.collectionId = id != null ? id : name;
        } catch (RestClientException e) {
            throw new RuntimeException("Cannot connect to ChromaDB", e);
        }
    }

    @Override
    public void addDocuments(List<String> documents, List<float[]> embeddings, List<Map<String, Object>> metadata) {
        List<String> ids = new ArrayList<>();
        List<String> docs = new ArrayList<>();
        List<List<Double>> embs = new ArrayList<>();
        List<Map<String, Object>> metas = new ArrayList<>();
        for (int i = 0; i < documents.size(); i++) {
            ids.add("doc_" + i);
            docs.add(documents.get(i));
            embs.add(toDoubles(embeddings.get(i)));
            metas.add(metadata.get(i));
        }
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("ids", ids);
        payload.put("documents", docs);
        payload.put("embeddings", embs);
        payload.put("metadatas", metas);
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        try {
            restTemplate.postForEntity(chromaUrl + "/api/v2/collections/" + collectionId + "/add", new HttpEntity<>(payload, headers), String.class);
        } catch (RestClientException e) {
            throw new RuntimeException("Failed to add documents to ChromaDB", e);
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<SearchResult> search(float[] queryEmbedding, int nResults) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("query_embeddings", List.of(toDoubles(queryEmbedding)));
        payload.put("n_results", nResults);
        payload.put("include", List.of("documents", "metadatas", "distances"));
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        try {
            ResponseEntity<Map> res = restTemplate.postForEntity(chromaUrl + "/api/v2/collections/" + collectionId + "/query", new HttpEntity<>(payload, headers), Map.class);
            Map<String, Object> body = res.getBody();
            if (body == null) return new ArrayList<>();
            List<List<Object>> docs = (List<List<Object>>) body.getOrDefault("documents", List.of(List.of()));
            List<List<Object>> metas = (List<List<Object>>) body.getOrDefault("metadatas", List.of(List.of()));
            List<List<Object>> dists = (List<List<Object>>) body.getOrDefault("distances", List.of(List.of()));
            List<Object> d0 = docs.isEmpty() ? List.of() : docs.get(0);
            List<Object> m0 = metas.isEmpty() ? List.of() : metas.get(0);
            List<Object> dis0 = dists.isEmpty() ? List.of() : dists.get(0);
            List<SearchResult> results = new ArrayList<>();
            int size = Math.min(d0.size(), Math.min(m0.size(), dis0.size()));
            for (int i = 0; i < size; i++) {
                SearchResult r = new SearchResult();
                r.setDocument(d0.get(i) != null ? d0.get(i).toString() : "");
                r.setMetadata(m0.get(i) instanceof Map ? (Map<String, Object>) m0.get(i) : Map.of());
                r.setDistance(((Number) dis0.get(i)).doubleValue());
                results.add(r);
            }
            return results;
        } catch (RestClientException e) {
            throw new RuntimeException("Failed to query ChromaDB", e);
        }
    }

    @Override
    public void deleteCollection(String collectionName) {
        try {
            restTemplate.delete(chromaUrl + "/api/v2/collections/" + collectionName);
        } catch (RestClientException e) {
            // ignore
        }
    }

    @Override
    public List<String> listCollections() {
        try {
            ResponseEntity<List> res = restTemplate.getForEntity(chromaUrl + "/api/v2/collections", List.class);
            List<?> body = res.getBody();
            if (body == null) return new ArrayList<>();
            List<String> names = new ArrayList<>();
            for (Object item : body) {
                if (item instanceof Map) {
                    Object nameVal = ((Map<?, ?>) item).get("name");
                    names.add(nameVal == null ? "" : nameVal.toString());
                }
            }
            return names;
        } catch (RestClientException e) {
            return new ArrayList<>();
        }
    }

    @Override
    public void close() {
        this.collectionId = null;
    }

    private static List<Double> toDoubles(float[] arr) {
        List<Double> list = new ArrayList<>();
        for (float f : arr) {
            list.add((double) f);
        }
        return list;
    }
}

File: SemanticSearch/Elixir/Web/Phoenix/src/lib/semantic_search/vectorstore/chroma_db_vector_store.ex
defmodule SemanticSearch.VectorStore.ChromaDBVectorStore do
  @moduledoc """
  ChromaDB vector store over its HTTP API (v2). Uses `:req` for HTTP calls.
  """

  @behaviour SemanticSearch.VectorStore.VectorStoreAdapter

  alias SemanticSearch.VectorStore.SearchResult

  @impl true
  def connect do
    url = chroma_url()

    case Req.get!(url <> "/api/v2/collections", receive_timeout: 2000) do
      %Req.Response{status: s} when s in 200..299 ->
        {:ok, :connected}

      %Req.Response{} ->
        {:error, :cannot_connect}
    end
  rescue
    _ -> {:error, :cannot_connect}
  end

  @impl true
  def add_documents(documents, embeddings, metadata) do
    collection_id = get_or_create_collection()

    payload = %{
      ids: Enum.with_index(documents) |> Enum.map(fn {_, i} -> "doc_#{i}" end),
      documents: documents,
      embeddings: embeddings,
      metadatas: metadata
    }

    url = chroma_url() <> "/api/v2/collections/#{collection_id}/add"

    case Req.post!(url, json: payload, receive_timeout: 5000) do
      %Req.Response{status: s} when s in 200..299 -> :ok
      %Req.Response{} -> {:error, :add_failed}
    end
  rescue
    _ -> {:error, :add_failed}
  end

  @impl true
  def search(query_embedding, n_results) do
    collection_id = get_or_create_collection()

    payload = %{
      query_embeddings: [query_embedding],
      n_results: n_results,
      include: ["documents", "metadatas", "distances"]
    }

    url = chroma_url() <> "/api/v2/collections/#{collection_id}/query"

    case Req.post!(url, json: payload, receive_timeout: 5000) do
      %Req.Response{status: s, body: body} when s in 200..299 ->
        docs = List.wrap(get_in(body, ["documents", Access.at(0)]) || [])
        metas = List.wrap(get_in(body, ["metadatas", Access.at(0)]) || [])
        dists = List.wrap(get_in(body, ["distances", Access.at(0)]) || [])

        docs
        |> Enum.zip([metas, dists] |> Enum.map(&List.wrap/1))
        |> Enum.map(fn {doc, {meta, dist}} ->
          %SearchResult{
            document: to_string(doc),
            metadata: if(is_map(meta), do: meta, else: %{}),
            distance: if(is_number(dist), do: dist, else: 0.0)
          }
        end)

      %Req.Response{} ->
        []
    end
  rescue
    _ -> []
  end

  @impl true
  def delete_collection(name) do
    url = chroma_url() <> "/api/v2/collections/#{name}"

    try do
      Req.delete!(url, receive_timeout: 5000)
      :ok
    rescue
      _ -> {:error, :delete_failed}
    end
  end

  @impl true
  def list_collections do
    url = chroma_url() <> "/api/v2/collections"

    case Req.get!(url, receive_timeout: 5000) do
      %Req.Response{status: s, body: body} when s in 200..299 ->
        Enum.map(List.wrap(body), & &1["name"])

      %Req.Response{} ->
        []
    end
  rescue
    _ -> []
  end

  defp get_or_create_collection do
    collection = System.get_env("VECTOR_COLLECTION", "documents")
    existing = find_collection(collection)
    if existing, do: existing, else: create_collection(collection)
  end

  defp find_collection(name) do
    url = chroma_url() <> "/api/v2/collections"

    case Req.get!(url, receive_timeout: 5000) do
      %Req.Response{status: s, body: body} when s in 200..299 ->
        Enum.find_value(List.wrap(body), fn c ->
          if c["name"] == name, do: c["id"], else: nil
        end)

      %Req.Response{} ->
        nil
    end
  rescue
    _ -> nil
  end

  defp create_collection(name) do
    payload = %{
      name: name,
      metadata: %{"hnsw:space" => "cosine"},
      get_or_create: true
    }

    url = chroma_url() <> "/api/v2/collections"

    case Req.post!(url, json: payload, receive_timeout: 5000) do
      %Req.Response{status: s, body: body} when s in 200..299 ->
        body["id"] || body["name"] || name

      %Req.Response{} ->
        name
    end
  rescue
    _ -> name
  end

  defp chroma_url do
    case System.get_env("CHROMA_URL", "") do
      "" ->
        host = System.get_env("CHROMA_HOST", "localhost")
        port = System.get_env("CHROMA_PORT", "8000")
        "http://#{host}:#{port}"

      url ->
        url |> String.trim_trailing("/")
    end
  end
end
